### PR TITLE
[ServiceBus] Dynamic link-credit issuing when calling receive

### PR DESCRIPTION
**Pull request by yunhaoling**

### CODE REVIEW

**Comment by yunhaoling** [comment]
pending uamqp v1.2.9 release

File: sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py [hunk 1]
```diff
@@ -182,14 +182,45 @@ def _open(self):
...
+                if abs_timeout_ms and amqp_receive_client._counter.get_current_ms() > abs_timeout_ms:
+                    expired = True
+                    break
+                before = received_messages_queue.qsize()
+                receiving = amqp_receive_client.do_work()
+                received = received_messages_queue.qsize() - before
+                if not first_message_received and received_messages_queue.qsize() > 0 and received > 0:
+                    # first message(s) received, continue receiving for some time
+                    first_message_received = True
+                    abs_timeout_ms = amqp_receive_client._counter.get_current_ms() + \
```
Comment (KieranBrantnerMagee) [inline sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py:219]: Just double checking that we never refresh this timeout as more come in; it's only "time after first"?

File: sdk/servicebus/azure-servicebus/samples/sync_samples/receive_queue.py [hunk 2]
```diff
@@ -20,7 +20,7 @@
 servicebus_client = ServiceBusClient.from_connection_string(conn_str=CONNECTION_STR)
 
 with servicebus_client:
-    receiver = servicebus_client.get_queue_receiver(queue_name=QUEUE_NAME, prefetch=10)
+    receiver = servicebus_client.get_queue_receiver(queue_name=QUEUE_NAME)
```
Comment (KieranBrantnerMagee) [inline sdk/servicebus/azure-servicebus/samples/sync_samples/receive_queue.py:23]: Do we use or discuss prefetch in the README anywhere that we'd want to nix?

File: sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py [hunk 3]
```diff
@@ -182,14 +182,45 @@ def _open(self):
...
+        while not received_messages_queue.empty() and len(batch) < max_batch_size:
+            batch.append(received_messages_queue.get())
+            received_messages_queue.task_done()
+        if len(batch) >= max_batch_size:
+            return batch
+
+        # Dynamically issue link credit if max_batch_size > 1 when the prefetch is the default value 1
+        if max_batch_size and self._prefetch == 1 and max_batch_size > 1:
+            link_credit_needed = max_batch_size - len(batch)
+            amqp_receive_client.message_handler.reset_link_credit(link_credit_needed)
```
Comment (annatisch) [inline sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py:204]: Do we need to check for a max/min link credit value before calling this?

Comment (annatisch) [reply sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py:204]: (I'm curious what happens if this gets passed "-1" or "100000"

File: sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py [hunk 4]
```diff
@@ -309,9 +340,6 @@ def receive(self, max_batch_size=None, max_wait_time=None):
 
         """
         self._check_live()
-        if max_batch_size and self._prefetch < max_batch_size:
```
Comment (annatisch) [inline sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py:312]: Do we need to update the documentation of the `prefetch` parameter?

File: sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py (hunk 3 above)
Comment (yunhaoling) [reply sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py:204]: I don't think we need.

for the min (negative value) case: the prior comparison between `len(batch)` and `max_batch_size` guarantees that `link_credit_needed` is > 0. (and the underlying C code only accepts `uint` type for link credit`

for the max (very big value) case: I don't see limitation from the service doc. I've tested EventHub, the link credit can be as large as the `uint` max -- 4294967295.

File: sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py (hunk 4 above)
Comment (yunhaoling) [reply sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py:312]: yes, I'll do the update

File: sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py (hunk 1 above)
Comment (yunhaoling) [reply sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py:219]: correct! only the first time

File: sdk/servicebus/azure-servicebus/samples/sync_samples/receive_queue.py (hunk 2 above)
Comment (yunhaoling) [reply sdk/servicebus/azure-servicebus/samples/sync_samples/receive_queue.py:23]: README doesn't mention "prefetch".

But I have updated the docstring and samples.

File: sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py (hunk 4 above)
Comment (yunhaoling) [reply sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py:312]: updated, add extra explanation for the behavior the receive when 'prefetch' is 0

File: sdk/servicebus/azure-servicebus/samples/sync_samples/receive_queue.py (hunk 2 above)
Comment (KieranBrantnerMagee) [reply sdk/servicebus/azure-servicebus/samples/sync_samples/receive_queue.py:23]: Echoing from our OOB chat for posterity:

I might just rearrange your stanzas.

 

Prefetch: the maximum <etc etc.> to the service.
This setting is only for advanced performance tuning, Increasing this value will <the rest of your warning here>
Default value is 0, meaning <etc>; in the case of 0 <etc>

(Feel free to ping me whenever you're green on tests and I can OK this)

Comment (yunhaoling) [reply sdk/servicebus/azure-servicebus/samples/sync_samples/receive_queue.py:23]: already updated, thanks for the suggestion!

**Pull request merged**